### PR TITLE
Feat/218 better error handling

### LINKED ISSUE #218: UX: better handling of unavailable backend
|</img> @dsimansk | [2017-03-23](https://github.com/syndesisio/syndesis-ui/issues/307) | enhancement, Priority - Low, UX Design  |
|-|-|-|

Currently if ipaas-rest is not responding for some reason (misconfig of auth, 503 pod not available etc.) There's just spinner spinning endlessly. We should probably have timeout to request and provide laundered error message to end user, that something is terribly wrong.

### LINKED ISSUE #2315: Incorrect error message displayed when uploading an invalid swagger file
## This is a...

<pre><code>
[ ] Feature request
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Documentation issue or request
</code></pre>

## The problem

When uploading an invalid swagger file it appears the API can either return an HTTP 200 with an `error` attribute *or* return an HTTP 500 error response with the following:

```
{"errorCode":500,"userMsg":"Please contact the administrator and file a bug report","developerMsg":"Internal Server Exception. Swagger specification does not provide a `schemes` definition and the Swagger specification was uploaded so the originating URL is lost to determine the scheme to use"}
```

In the latter case where the above response is received the wrong error message is shown in the UI:

screenshot 2018-04-10 09 52 58

Probably caught by a generic error handler and not correctly propagated back to the component.

This can be reproduced by trying to create an API client connector with this swagger file:

[kie-server-swagger.json.txt](https://github.com/syndesisio/syndesis/files/1894902/kie-server-swagger.json.txt)

@chirino FYI

### PR DESCRIPTION
**Pull request by deeleman**
Fixes #218 
Fixes #2315 too

This PR encompass the following new features:

- Application wide errors are now funneled through a centralized Exception Handler Service, allowing for easier logging and reporting.
- Http errors (HTTP 500 errors, API calls yielding a 404 error, etc) when communicating with the backend API are now handled and not propagated, ensuring that components keep being responsive, and also reported through a toast notification. This will enhance the overall UX on those (many) components/flows where no proper Http exception handling had been implemented.
- Network errors (WiFi lags, Internet downtimes) are now reported on realtime with a toast notification informing the user that his Internet connection has been interrupted or is back.

FYI, the error strings configured for each toast notification are fully managed through the i18n dictionary.

### CODE REVIEW

File: app/ui/src/app/api/providers/api-http-provider.service.ts [hunk 1]
```diff
@@ -150,7 +151,7 @@ export class ApiHttpProviderService extends ApiHttpService {
       .filter(requestEvent => requestEvent.type === HttpEventType.Response)
       .map(requestEvent => requestEvent as HttpResponse<T>)
       .map(requestEvent => requestEvent.body)
-      .catch(error => Observable.throw(this.catchError(JSON.parse(error.error))));
+      .catch(error => this.handleError(error.error));
```
Comment (gashcrumb) [inline app/ui/src/app/api/providers/api-http-provider.service.ts]: I'd suggest passing the whole error object here too...

Comment (gashcrumb) [reply app/ui/src/app/api/providers/api-http-provider.service.ts]: This is basically the change I had in #2342 let's just go with this PR for that.

Comment (deeleman) [reply app/ui/src/app/api/providers/api-http-provider.service.ts]: Problem is that, contrary to the other `httpClient` handlers, the object returned through the RxJS `catch` handler here is not the actual HTTP error, but a native Angular wrapper around it, typed as a `HttpErrorResponse` object. Therefore the properties will not match with those expected by the logic in `handleError()`.

More info: https://github.com/angular/angular/blob/5.2.9/packages/common/http/src/response.ts#L288-L329

Comment (gashcrumb) [reply app/ui/src/app/api/providers/api-http-provider.service.ts]: ah, okay gotcha didn't realize that, thanks!  Well then I'm all +1 on this PR then :-)

**Pull request merged**